Fix bugs in Tensor and TaggedTuple

nilsdeppe opened the pull request:
## Proposed changes

- Fix bug in Tensor
- Fix bug in TaggedTuple

### Types of changes:

- [x] Bugfix
- [ ] New feature

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] Follows [code review guidelines](https://sxs-collaboration.github.io/spectre/code_review_guide.html)
- [ ] Code has documentation and unit tests
- [ ] Private member variables have a trailing underscore
- [ ] Do not use [Hungarian notation](https://en.wikipedia.org/wiki/Hungarian_notation), e.g. `double* pd_blah` is bad
- [ ] Header order:
  1. hpp corresponding to cpp (only in cpp files)
  2. Blank line (only in cpp files)
  3. STL and externals (in alphabetical order)
  4. Blank line
  5. SpECTRE includes (in alphabetical order)
- [ ] File lists in CMake are alphabetical
- [ ] Correct `noexcept` specification for functions (if unsure, mark `noexcept`)
- [ ] Mark objects `const` whenever possible
- [ ] Almost always `auto`, except with expression templates, i.e. `DataVector`
- [ ] All commits for performance changes provide quantitative evidence and the tests used to obtain said evidence.
- [ ] Make sure error messages are helpful, e.g. "The number of grid points in the matrix 'F' is not the same as the number of grid points in the determinant."
- [ ] Prefix commits addressing PR requests with `fixup`


### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...
-->

nilsdeppe committed:
Make Tensor SFINAE-friendly

wthrowe commented:
"Fix bug in X" is not a very useful commit message.  I've looked at the diff and I still have no idea what the bug was beyond that it had something to do with one-element tuples (based on the test name).

Since this touches the code, under what circumstances are TaggedTuples supposed to be implicitly interconvertible?  It does not appear to be documented, and it's hard to decide if the change is reasonable when I don't know the desired behavior.

nilsdeppe committed:
Fix bugs TaggedTuple constructors

There are two issues, one dealing with correct reference type-ness in
the implementation, the other dealing with single-argument
construction of a single-tag TaggedTuple.

1. The TaggedTupleLeaf class was incorrectly checking if the
value_type is constructible from an rvalue reference rather than the
reference it received.

2. When move constructing a TaggedTuple<Tag1> from a TaggedTuple<Tag0>
(where Tag1::type is the same as Tag0::type) the necessary constructor
was not enable for overload resolution. The constructor of
TaggedTuple<TTags...> taking TaggedTuple<UTags...>&& should be
available for overload resolution when:
- sizeof...(TTags) == sizeof...(UTags)
- is_constructible_v(TTags::type, UTags::type&&) is true
- is_same_v<TTags, UTags> is false for at least one of TTags and UTags

The same rules apply for the const& constructor except rvalues are
replaced with const lvalues references.

nilsdeppe commented:
I've pushed an updated that addresses Will's comment. I've decided that no implicit conversion between TaggedTuple's with different tags but the same tag types should be allowed. Since this is effectively changing the keys in a map I want it to be done explicitly. This doesn't change any of our code that uses the TaggedTuple, but does simplify a lot of the logic inside TaggedTuple
